storagemarket/impl/providerstates/safereadercloser.go

```diff
@@ -0,0 +1,40 @@
+package providerstates
+
+import (
+	"io"
+	"sync"
+
+	"golang.org/x/xerrors"
+)
+
+type safeReaderCloser struct {
+	r io.Reader
+	c io.Closer
+
+	lk       sync.Mutex
+	isClosed bool
```

Comment:
You can simplify this to an atomic int32 and use the CAS (compare and swap primitives).